shadowsocks/stream.go

```diff
@@ -165,76 +166,91 @@ func (sr *shadowsocksReader) init() (err error) {
+// Returns an error only if the block could not be read.
+func (sr *shadowsocksReader) readBlock(buf []byte) error {
 	_, err := io.ReadFull(sr.reader, buf)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	buf, err = sr.aead.Open(buf[:0], sr.counter, buf, nil)
+	_, err = sr.aead.Open(buf[:0], sr.counter, buf, nil)
 	increment(sr.counter)
 	if err != nil {
-		return nil, fmt.Errorf("failed to decrypt: %v", err)
+		return fmt.Errorf("failed to decrypt: %v", err)
 	}
-	return buf, nil
+	return nil
 }
 
-func (sr *shadowsocksReader) Read(b []byte) (int, error) {
-	if err := sr.ensureLeftover(); err != nil {
+// ReadChunk returns the next decrypted chunk.  The caller must
+// complete its use of the returned buffer before the next call.
+func (sr *shadowsocksReader) ReadChunk() ([]byte, error) {
+	if err := sr.init(); err != nil {
+		return nil, err
+	}
+	sizeBuf := sr.buf[:2+sr.aead.Overhead()]
+	if err := sr.readBlock(sizeBuf); err != nil {
+		if err != io.EOF && err != io.ErrUnexpectedEOF {
+			err = fmt.Errorf("failed to read payload size: %v", err)
+		}
+		return nil, err
+	}
+	size := int(binary.BigEndian.Uint16(sizeBuf) & payloadSizeMask)
+	payloadBuf := sr.buf[:size+sr.aead.Overhead()]
+	if err := sr.readBlock(payloadBuf); err != nil {
+		if err == io.EOF { // EOF is not expected mid-chunk.
+			err = io.ErrUnexpectedEOF
+		}
+		return nil, err
+	}
+	return payloadBuf[:size], nil
+}
+
+// readConverter adapts from shadowsocksReader, with source-controlled
+// chunk sizes, to Go-style IO.
+type readConverter struct {
+	r        *shadowsocksReader
+	leftover []byte
+}
+
+func (c *readConverter) Read(b []byte) (int, error) {
+	if err := c.ensureLeftover(); err != nil {
 		return 0, err
 	}
-	n := copy(b, sr.leftover)
-	sr.leftover = sr.leftover[n:]
+	n := copy(b, c.leftover)
+	c.leftover = c.leftover[n:]
 	return n, nil
 }
 
-func (sr *shadowsocksReader) WriteTo(w io.Writer) (written int64, err error) {
+func (c *readConverter) WriteTo(w io.Writer) (written int64, err error) {
 	for {
-		if err = sr.ensureLeftover(); err != nil {
+		if err = c.ensureLeftover(); err != nil {
 			if err == io.EOF {
 				err = nil
 			}
 			return written, err
 		}
-		n, err := w.Write(sr.leftover)
+		n, err := w.Write(c.leftover)
 		written += int64(n)
-		sr.leftover = sr.leftover[n:]
+		c.leftover = c.leftover[n:]
 		if err != nil {
 			return written, err
 		}
 	}
 }
 
-// Ensures that sr.leftover is nonempty.  If leftover is empty, this method
+// Ensures that c.leftover is nonempty.  If leftover is empty, this method
 // waits for incoming data and decrypts it.
-// Returns an error only if sr.leftover could not be populated.
-func (sr *shadowsocksReader) ensureLeftover() error {
-	if len(sr.leftover) > 0 {
+// Returns an error only if c.leftover could not be populated.
+func (c *readConverter) ensureLeftover() error {
+	if len(c.leftover) > 0 {
 		return nil
 	}
-	buf, err := sr.readBlock(2)
+	payload, err := c.r.ReadChunk()
```

Comment:
payload -> chunk to be consistent

Reply:
I've adjusted the names to try to match the Shadowsocks AEAD spec page on shadowsocks.org, and added a comment.  I think there is a distinction between "chunk" (wire format) and "payload" (plaintext contents).